File: Interface utilisateur /Moteur.java
package com.quoridor;

// Importation des structures de données dynamiques pour la gestion des collections et des files d'attente
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

// Moteur central gérant la logique mathématique et les règles strictes du jeu Quoridor
public class Moteur {
    
    // Constante définissant la dimension standard de la matrice de jeu
    public final int NB_CASES = 9;
    
    // Coordonnées cartésiennes initiales et réserve d'obstacles pour le joueur humain
    public int pionBlancLigne = 8;
    public int pionBlancCol = 4;
    public int mursJoueur = 10;
    
    // Coordonnées cartésiennes initiales et réserve d'obstacles pour l'Intelligence Artificielle
    public int pionNoirLigne = 0; 
    public int pionNoirCol = 4;   
    public int mursIA = 10; 

    // Indicateur booléen signalant la conclusion mathématique de l'affrontement
    public boolean partieTerminee = false;
    
    // Structure de données mémorisant l'historique des obstacles physiquement déployés sur le plateau
    public List<MurLogique> mursPlaques = new ArrayList<>();

    // Classe interne modélisant la structure logique et spatiale d'un mur
    public class MurLogique {
        public int ligne;
        public int col;
        public boolean horizontal;
        
        // Constructeur initialisant l'obstacle avec ses coordonnées et son orientation spatiale
        public MurLogique(int l, int c, boolean h) { this.ligne = l; this.col = c; this.horizontal = h; }
    }

    // Évalue géométriquement si la transition entre deux cellules adjacentes est obstruée par un mur
    public boolean cheminLibreDeMurs(int ligne1, int col1, int ligne2, int col2) {
        
        // Itération sur l'ensemble des obstacles actifs pour détecter une éventuelle collision spatiale
        for (MurLogique mur : mursPlaques) {
            
            // Détection d'une obstruction géométrique lors d'un déplacement vers le nord
            if (ligne2 < ligne1 && mur.horizontal && mur.ligne == ligne2 && (mur.col == col1 || mur.col == col1 - 1)) return false;
            
            // Détection d'une obstruction géométrique lors d'un déplacement vers le sud
            if (ligne2 > ligne1 && mur.horizontal && mur.ligne == ligne1 && (mur.col == col1 || mur.col == col1 - 1)) return false;
            
            // Détection d'une obstruction géométrique lors d'un déplacement vers l'est
            if (col2 > col1 && !mur.horizontal && mur.col == col1 && (mur.ligne == ligne1 || mur.ligne == ligne1 - 1)) return false;
            
            // Détection d'une obstruction géométrique lors d'un déplacement vers l'ouest
            if (col2 < col1 && !mur.horizontal && mur.col == col2 && (mur.ligne == ligne1 || mur.ligne == ligne1 - 1)) return false;
        }
        return true;
    }

    // Implémentation de l'algorithme BFS pour certifier l'existence d'un graphe de cheminement valide
    public boolean aUnChemin(int ligneDepart, int colDepart, int ligneCible) {
        
        // Matrice booléenne de mémorisation spatiale pour éviter le traitement redondant de l'algorithme BFS
        boolean[][] visite = new boolean[NB_CASES][NB_CASES];
        
        // Structure de type FIFO régulant l'ordre d'exploration géographique des nœuds
        Queue<int[]> fileAttente = new LinkedList<>();
        
        fileAttente.add(new int[]{ligneDepart, colDepart});
        visite[ligneDepart][colDepart] = true;
        
        // Vecteurs de déplacement standardisés modélisant les mouvements orthogonaux
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        // Boucle de propagation explorant la grille tant qu'il reste des cellules accessibles
        while (!fileAttente.isEmpty()) {
            int[] caseActuelle = fileAttente.poll();
            int ligne = caseActuelle[0];
            int col = caseActuelle[1];

            // Interruption de la recherche asynchrone si les coordonnées atteignent la cible stratégique
            if (ligne == ligneCible) return true;

            // Évaluation systématique des cellules adjacentes selon les vecteurs directionnels
            for (int[] direction : directions) {
                int prochaineLigne = ligne + direction[0];
                int prochaineCol = col + direction[1];

                // Intégration du nœud voisin dans la file d'attente s'il respecte les contraintes physiques
                if (prochaineLigne >= 0 && prochaineLigne < NB_CASES && prochaineCol >= 0 && prochaineCol < NB_CASES) {
                    if (!visite[prochaineLigne][prochaineCol] && cheminLibreDeMurs(ligne, col, prochaineLigne, prochaineCol)) {
                        visite[prochaineLigne][prochaineCol] = true;
                        fileAttente.add(new int[]{prochaineLigne, prochaineCol});
                    }
                }
            }
        }
        
        // Retourne un statut d'échec si l'algorithme BFS a épuisé la totalité des possibilités d'exploration
        return false; 
    }

    // Valide mathématiquement et topologiquement l'autorisation de déployer un nouvel obstacle
    public boolean murEstValide(int ligne, int col, boolean estHorizontal) {
        
        // Analyse préventive des superpositions spatiales avec les structures déjà implémentées
        for (MurLogique mur : mursPlaques) {
            
            // Identification d'un conflit de coordonnées strict
            if (mur.ligne == ligne && mur.col == col) return false;
            
            // Identification d'une collision géométrique sur l'axe horizontal
            if (estHorizontal && mur.horizontal && mur.ligne == ligne && (mur.col == col - 1 || mur.col == col + 1)) return false;
            
            // Identification d'une collision géométrique sur l'axe vertical
            if (!estHorizontal && !mur.horizontal && mur.col == col && (mur.ligne == ligne - 1 || mur.ligne == ligne + 1)) return false;
        }

        // Déploiement temporaire de la structure en mémoire pour simuler ses conséquences sur le graphe de jeu
        MurLogique murTest = new MurLogique(ligne, col, estHorizontal);
        mursPlaques.add(murTest);

        // Invocation du contrôle BFS pour garantir le respect strict de la règle d'anti-enfermement
        boolean blancPeutFinir = aUnChemin(pionBlancLigne, pionBlancCol, 0);
        boolean noirPeutFinir = aUnChemin(pionNoirLigne, pionNoirCol, 8);

        // Suppression de l'obstacle simulé après la validation algorithmique
        mursPlaques.remove(mursPlaques.size() - 1);
        
        // Autorise la modification de l'environnement si les deux entités conservent un chemin vers la victoire
        return blancPeutFinir && noirPeutFinir;
    }

    // Analyse géométrique et logique déterminant la validité d'une translation de pion
    public boolean estDeplacementValide(int ligneDepart, int colDepart, int ligneArrivee, int colArrivee) {
        
        // Extraction des coordonnées cartésiennes de l'entité adverse pour anticiper les collisions
        int ligneAdversaire = (ligneDepart == pionBlancLigne && colDepart == pionBlancCol) ? pionNoirLigne : pionBlancLigne;
        int colAdversaire = (ligneDepart == pionBlancLigne && colDepart == pionBlancCol) ? pionNoirCol : pionBlancCol;

        // Modélisation du déplacement orthogonal standard d'une unité de distance
        if (Math.abs(ligneDepart - ligneArrivee) + Math.abs(colDepart - colArrivee) == 1) {
            
            // Rejet de la translation si la cellule cible est occupée par l'adversaire
            if (ligneArrivee == ligneAdversaire && colArrivee == colAdversaire) return false; 
            
            // Autorisation finale sous réserve d'absence de barrière architecturale sur le segment
            return cheminLibreDeMurs(ligneDepart, colDepart, ligneArrivee, colArrivee);
        }

        // Évaluation de la trajectoire pour un saut acrobatique sur l'axe vertical
        if (Math.abs(ligneDepart - ligneArrivee) == 2 && colDepart == colArrivee) {
            int ligneMilieu = (ligneDepart + ligneArrivee) / 2;
            if (ligneMilieu == ligneAdversaire && colDepart == colAdversaire) { 
                return cheminLibreDeMurs(ligneDepart, colDepart, ligneMilieu, colDepart) && cheminLibreDeMurs(ligneMilieu, colDepart, ligneArrivee, colArrivee);
            }
        }
        
        // Évaluation de la trajectoire pour un saut acrobatique sur l'axe horizontal
        if (Math.abs(colDepart - colArrivee) == 2 && ligneDepart == ligneArrivee) {
            int colMilieu = (colDepart + colArrivee) / 2;
            if (ligneDepart == ligneAdversaire && colMilieu == colAdversaire) {
                return cheminLibreDeMurs(ligneDepart, colDepart, ligneDepart, colMilieu) && cheminLibreDeMurs(ligneDepart, colMilieu, ligneArrivee, colArrivee);
            }
        }

        // Traitement d'exception mathématique autorisant un déplacement diagonal sous contraintes
        if (Math.abs(ligneDepart - ligneArrivee) == 1 && Math.abs(colDepart - colArrivee) == 1) {
            
            // Analyse du scénario géométrique où l'entité bloquante est alignée sur l'axe horizontal
            if (ligneAdversaire == ligneDepart && Math.abs(colAdversaire - colDepart) == 1) {
                int colArriere = colAdversaire + (colAdversaire - colDepart); 
                
                // Détection de l'impossibilité physique d'effectuer un saut standard due à la bordure ou à un mur
                boolean estBloque = (colArriere < 0 || colArriere >= NB_CASES || !cheminLibreDeMurs(ligneAdversaire, colAdversaire, ligneAdversaire, colArriere));
                
                // Validation du chemin de contournement diagonal si l'axe principal s'avère totalement obstrué
                if (estBloque && colArrivee == colAdversaire) {
                    return cheminLibreDeMurs(ligneDepart, colDepart, ligneAdversaire, colAdversaire) && cheminLibreDeMurs(ligneAdversaire, colAdversaire, ligneArrivee, colArrivee);
                }
            }
            // Analyse du scénario géométrique où l'entité bloquante est alignée sur l'axe vertical
            else if (colAdversaire == colDepart && Math.abs(ligneAdversaire - ligneDepart) == 1) {
                int ligneArriere = ligneAdversaire + (ligneAdversaire - ligneDepart); 
                
                // Détection de l'impossibilité physique d'effectuer un saut standard due à la bordure ou à un mur
                boolean estBloque = (ligneArriere < 0 || ligneArriere >= NB_CASES || !cheminLibreDeMurs(ligneAdversaire, colAdversaire, ligneArriere, colAdversaire));
                
                // Validation du chemin de contournement diagonal si l'axe principal s'avère totalement obstrué
                if (estBloque && ligneArrivee == ligneAdversaire) {
                    return cheminLibreDeMurs(ligneDepart, colDepart, ligneAdversaire, colAdversaire) && cheminLibreDeMurs(ligneAdversaire, colAdversaire, ligneArrivee, colArrivee);
                }
            }
        }
        
        // Rejet définitif de la requête si la signature géométrique ne correspond à aucun modèle légal du système
        return false;
    }
}

File: Intelligence Artificielle python/moteur.py
# Importation de la structure de données à double extrémité pour l'optimisation des parcours de graphes
from collections import deque

# Importation du module stochastique pour l'échantillonnage aléatoire des actions
import random

# Structure de données encapsulant l'état topologique et matériel de la grille de Quoridor
class EtatQuoridor:
    
    def __init__(self):
        # Instanciation des coordonnées cartésiennes initiales pour l'Intelligence Artificielle et le joueur humain
        self.position_ia = (0, 4)      
        self.position_joueur = (8, 4)  
        
        # Allocation de l'inventaire matériel limitant les capacités d'obstruction
        self.murs_ia = 10
        self.murs_joueur = 10
        
        # Ensembles de hachage mémorisant les coordonnées spatiales des obstacles déployés
        self.murs_horizontaux = set()
        self.murs_verticaux = set()

# Évaluation géométrique certifiant l'absence d'interférence physique entre deux nœuds adjacents
def est_chemin_libre(etat, ligne1, col1, ligne2, col2):
    # Détection des collisions potentielles sur l'axe des ordonnées induites par des structures horizontales
    if ligne2 < ligne1: return not ((ligne2, col1) in etat.murs_horizontaux or (ligne2, col1 - 1) in etat.murs_horizontaux) 
    if ligne2 > ligne1: return not ((ligne1, col1) in etat.murs_horizontaux or (ligne1, col1 - 1) in etat.murs_horizontaux) 
    
    # Détection des collisions potentielles sur l'axe des abscisses induites par des structures verticales
    if col2 > col1: return not ((ligne1, col1) in etat.murs_verticaux or (ligne1 - 1, col1) in etat.murs_verticaux)     
    if col2 < col1: return not ((ligne1, col2) in etat.murs_verticaux or (ligne1 - 1, col2) in etat.murs_verticaux)     
    
    return True

# Identification matricielle de tous les nœuds accessibles depuis une coordonnée source spécifique
def obtenir_voisins_valides(etat, position, est_ia=True, ignorer_joueurs=False):
    ligne, col = position
    position_adversaire = etat.position_joueur if est_ia else etat.position_ia
    voisins = []
    
    # Vecteurs de translation modélisant les mouvements orthogonaux stricts
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)] 
    
    # Itération sur l'espace vectoriel pour valider l'intégrité topologique des déplacements
    for delta_ligne, delta_col in directions:
        prochaine_ligne, prochaine_col = ligne + delta_ligne, col + delta_col
        
        # Vérification stricte de l'appartenance à la matrice spatiale et de l'absence d'obstruction
        if 0 <= prochaine_ligne <= 8 and 0 <= prochaine_col <= 8 and est_chemin_libre(etat, ligne, col, prochaine_ligne, prochaine_col):
            
            # Traitement du cas standard caractérisé par une cellule cible inoccupée
            if ignorer_joueurs or (prochaine_ligne, prochaine_col) != position_adversaire:
                voisins.append((prochaine_ligne, prochaine_col))
                
            # Application de la résolution mathématique de collision lors de la rencontre avec l'entité antagoniste
            elif not ignorer_joueurs and (prochaine_ligne, prochaine_col) == position_adversaire:
                
                saut_ligne, saut_col = prochaine_ligne + delta_ligne, prochaine_col + delta_col
                
                # Projection du saut acrobatique rectiligne par-dessus l'obstacle dynamique
                if 0 <= saut_ligne <= 8 and 0 <= saut_col <= 8 and est_chemin_libre(etat, prochaine_ligne, prochaine_col, saut_ligne, saut_col):
                    voisins.append((saut_ligne, saut_col))
                
                # Dérivation des trajectoires diagonales secondaires si le vecteur principal est géométriquement bloqué
                else:
                    if delta_ligne != 0: 
                        if col > 0 and est_chemin_libre(etat, prochaine_ligne, prochaine_col, prochaine_ligne, prochaine_col - 1): voisins.append((prochaine_ligne, prochaine_col - 1))
                        if col < 8 and est_chemin_libre(etat, prochaine_ligne, prochaine_col, prochaine_ligne, prochaine_col + 1): voisins.append((prochaine_ligne, prochaine_col + 1))
                    else: 
                        if ligne > 0 and est_chemin_libre(etat, prochaine_ligne, prochaine_col, prochaine_ligne - 1, prochaine_col): voisins.append((prochaine_ligne - 1, prochaine_col))
                        if ligne < 8 and est_chemin_libre(etat, prochaine_ligne, prochaine_col, prochaine_ligne + 1, prochaine_col): voisins.append((prochaine_ligne + 1, prochaine_col))
                        
    return voisins

# Résolution de la distance minimale absolue via l'algorithme BFS appliqué à la matrice de jeu
def obtenir_longueur_chemin_court(etat, position_depart, ligne_arrivee):
    # Initialisation de la file FIFO et du registre d'exploration pour prévenir la redondance cyclique
    file_a_explorer = deque([(position_depart, 0)]) 
    cases_visitees = set([position_depart])
    
    # Propagation itérative de la recherche topologique jusqu'à l'intersection avec l'axe d'arrivée
    while file_a_explorer:
        case_actuelle, distance = file_a_explorer.popleft()
        
        if case_actuelle[0] == ligne_arrivee:
            return distance 
            
        for voisin in obtenir_voisins_valides(etat, case_actuelle, ignorer_joueurs=True):
            if voisin not in cases_visitees:
                cases_visitees.add(voisin)
                file_a_explorer.append((voisin, distance + 1))
                
    # Retourne l'infini mathématique attestant d'un isolement complet de l'entité
    return float('inf') 

# Analyse structurelle validant l'insertion spatiale d'une barrière sans conflit topologique
def est_mur_valide(etat, ligne, col, orientation):
    if ligne < 0 or ligne > 7 or col < 0 or col > 7: return False
    
    # Prévention des chevauchements stricts et partiels pour une configuration de type horizontale
    if orientation == "H":
        if (ligne, col) in etat.murs_horizontaux or (ligne, col-1) in etat.murs_horizontaux or (ligne, col+1) in etat.murs_horizontaux or (ligne, col) in etat.murs_verticaux: return False
    # Prévention des chevauchements stricts et partiels pour une configuration de type verticale
    else:
        if (ligne, col) in etat.murs_verticaux or (ligne-1, col) in etat.murs_verticaux or (ligne+1, col) in etat.murs_verticaux or (ligne, col) in etat.murs_horizontaux: return False
        
    return True

# Génération exhaustive de l'espace des actions légales combinant translations et modifications structurelles
def obtenir_coups_legaux(etat, est_ia):
    coups = []
    position_actuelle = etat.position_ia if est_ia else etat.position_joueur
    murs_restants = etat.murs_ia if est_ia else etat.murs_joueur
    
    # Intégration systématique des vecteurs de mouvement mathématiquement autorisés
    for voisin in obtenir_voisins_valides(etat, position_actuelle, est_ia):
        coups.append(("MOVE", voisin[0], voisin[1]))
        
    # Évaluation conditionnelle de l'arsenal d'obstruction si le stock de l'entité le permet
    if murs_restants > 0:
        position_adversaire = etat.position_joueur if est_ia else etat.position_ia
        direction = -1 if est_ia else 1 
        
        # Réduction heuristique de l'espace de recherche ciblant le voisinage immédiat de l'entité adverse
        ligne_cible = position_adversaire[0] + direction
        if 0 <= ligne_cible <= 7:
            for colonne in range(max(0, position_adversaire[1] - 1), min(8, position_adversaire[1] + 1)):
                for orientation in ["H", "V"]:
                    
                    if est_mur_valide(etat, ligne_cible, colonne, orientation):
                        # Altération temporaire de l'espace de hachage pour anticiper les répercussions du placement
                        if orientation == "H": etat.murs_horizontaux.add((ligne_cible, colonne))
                        else: etat.murs_verticaux.add((ligne_cible, colonne))
                            
                        # Exécution de l'algorithme BFS garantissant l'absence d'enfermement mathématique définitif
                        if obtenir_longueur_chemin_court(etat, etat.position_ia, 8) != float('inf') and \
                           obtenir_longueur_chemin_court(etat, etat.position_joueur, 0) != float('inf'):
                            coups.append(("WALL", orientation, ligne_cible, colonne))
                            
                        # Restauration de l'intégrité de la mémoire spatiale après l'analyse de viabilité
                        if orientation == "H": etat.murs_horizontaux.remove((ligne_cible, colonne))
                        else: etat.murs_verticaux.remove((ligne_cible, colonne))
                        
    return coups

# Projection d'un état futur par l'application d'une action tout en mémorisant le point de retour
def simuler_coup(etat, coup, est_ia):
    position_annulation = etat.position_ia if est_ia else etat.position_joueur
    
    # Actualisation des coordonnées cartésiennes lors de l'exécution d'une translation
    if coup[0] == "MOVE":
        if est_ia: etat.position_ia = (coup[1], coup[2])
        else: etat.position_joueur = (coup[1], coup[2])
        
    # Décrémentation de l'inventaire matériel et enregistrement de la nouvelle coordonnée d'obstruction
    elif coup[0] == "WALL":
        if est_ia: etat.murs_ia -= 1
        else: etat.murs_joueur -= 1
        
        if coup[1] == "H": etat.murs_horizontaux.add((coup[2], coup[3]))
        else: etat.murs_verticaux.add((coup[2], coup[3]))
        
    return position_annulation

# Rétrogradation temporelle de l'état du graphe par réversion explicite de l'action simulée
def annuler_coup(etat, coup, position_annulation, est_ia):
    # Réassignation de la coordonnée spatiale originelle d'avant la translation
    if coup[0] == "MOVE":
        if est_ia: etat.position_ia = position_annulation
        else: etat.position_joueur = position_annulation
        
    # Restitution de l'actif matériel et extraction de la signature spatiale du système d'obstruction
    elif coup[0] == "WALL":
        if est_ia: etat.murs_ia += 1
        else: etat.murs_joueur += 1
        
        if coup[1] == "H": etat.murs_horizontaux.remove((coup[2], coup[3]))
        else: etat.murs_verticaux.remove((coup[2], coup[3])) 


# Clonage profond de l'objet pour préserver la structure lors de l'exploration de l'arbre des possibles
def copier_etat(etat):
    copie = EtatQuoridor()
    copie.position_ia = etat.position_ia
    copie.position_joueur = etat.position_joueur
    copie.murs_ia = etat.murs_ia
    copie.murs_joueur = etat.murs_joueur
    copie.murs_horizontaux = set(etat.murs_horizontaux)
    copie.murs_verticaux = set(etat.murs_verticaux)
    return copie

# Exécution d'une simulation de Monte-Carlo stochastique jusqu'à la détermination d'un état terminal
def jouer_partie_aleatoire(etat, tour_ia):
    # Instanciation d'un clone topologique pour isoler la simulation du flux temporel principal
    etat_simule = copier_etat(etat)
    est_ia = tour_ia
    
    # Itération de la boucle de résolution jusqu'à la validation mathématique d'un critère de victoire
    while etat_simule.position_ia[0] != 8 and etat_simule.position_joueur[0] != 0:
        # Extraction dynamique de l'arborescence des actions instantanées disponibles
        coups = obtenir_coups_legaux(etat_simule, est_ia)
        # Rupture de la boucle d'exécution si l'espace des solutions s'avère totalement vide
        if not coups:
            break
            
        # Sélection stochastique uniforme d'un vecteur d'action dans l'espace des possibles
        coup_choisi = random.choice(coups)
        # Application directe de la perturbation topologique issue de l'échantillonnage
        simuler_coup(etat_simule, coup_choisi, est_ia)
        # Inversion du flux de contrôle pour respecter l'alternance d'exécution asymétrique
        est_ia = not est_ia
        
    # Restitution d'un score de récompense binaire favorisant les occurrences de victoire de l'IA
    if etat_simule.position_ia[0] == 8:
        return 1
    return 0
